pkg/eventstream/eventstream.go

```diff
@@ -0,0 +1,98 @@
+package eventstream
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+var (
+	ErrInvalidRequestBody = fmt.Errorf("Request body must contain an event object or an array of event objects")
+	ErrEventTooLarge      = fmt.Errorf("Event is over the max size")
+)
+
+// MaxEvents is the maximum number of events we can parse in a single batch.
+const MaxEvents = 1_000
+
+// ParseStream parses a reader, publishing a stream of JSON-encoded events to the given channel,
+// ensuring that no individual event is too large.
+//
+// This closes the given channel once the JSON stream in the given reader has been parsed.
+//
+// Usage:
+//
+//      var err error
+//      go func() {
+//              err = ParseStream(ctx, r, stream)
+//      ()
+//
+//      for bytes := range stream {
+//              // consume event, transform event, etc
+//      }
+//
+//      if err != nil {
+//              // handle error
+//      }
+//
+func ParseStream(ctx context.Context, r io.Reader, stream chan json.RawMessage, maxSize int) error {
+	defer func() {
+		close(stream)
+	}()
+	d := json.NewDecoder(r)
+
+	token, err := d.Token()
+	if err == io.EOF {
+		return nil
+	}
+
+	delim, ok := token.(json.Delim)
+	if !ok {
+		// Invalid type
+		return ErrInvalidRequestBody
+	}
+
+	i := 0
+
+	switch delim {
```

Comment:
this is an interesting way of parsing JSON :thinking:

Reply:
A few things:
- We want to parse both arrays and objects.  This is hard with Go's crappy typing.
- We also want to parse as a stream, without allocating the entire thing up front, so we can parse each individual object of an array separately.

These requirements make this parsing... _unique_, to say the least :D